src/PointwiseFunctions/Hydro/EquationsOfState/Spectral.cpp

```diff
@@ -0,0 +1,309 @@
+  } else if constexpr (std::is_same_v<DataType, DataVector>) {
+    auto result = make_with_value<Scalar<DataVector>>(rest_mass_density, 0.0);
+    for (size_t i = 0; i < get(result).size(); ++i) {
+      get(result)[i] =
+          specific_internal_energy_from_density(get(rest_mass_density)[i]);
+    }
+    return result;
+  }
+}
+
+template <class DataType>
+Scalar<DataType> Spectral::chi_from_density_impl(
+    const Scalar<DataType>& rest_mass_density) const {
+  if constexpr (std::is_same_v<DataType, double>) {
+    return Scalar<double>{chi_from_density(get(rest_mass_density))};
+  } else if constexpr (std::is_same_v<DataType, DataVector>) {
+    auto result = make_with_value<Scalar<DataVector>>(rest_mass_density, 0.0);
+    for (size_t i = 0; i < get(result).size(); ++i) {
+      get(result)[i] = chi_from_density(get(rest_mass_density)[i]);
+    }
+    return result;
+  }
+}
+
+template <class DataType>
+Scalar<DataType> Spectral::kappa_times_p_over_rho_squared_from_density_impl(
+    const Scalar<DataType>& rest_mass_density) const {
+  return make_with_value<Scalar<DataType>>(get(rest_mass_density), 0.0);
+}
+
+// this evaluates the power series
+// Gamma(x) = Sum_{n=0}^N gamma_n x^n
+double Spectral::gamma(const double x) const {
+  return evaluate_polynomial(gamma_coefficients_, x);
+}
+
+double Spectral::chi_from_density(const double rest_mass_density) const {
+  const double x = log(rest_mass_density / reference_density_);
+  const double P = pressure_from_log_density(x);
+  const double chi = P / rest_mass_density *
+                     (x <= 0.0 ? integral_coefficients_[0]
+                               : (x < x_max_ ? gamma(x) : gamma(x_max_)));
+  return chi;
+}
+
+double Spectral::specific_internal_energy_from_density(
+    const double rest_mass_density) const {
+  const double x = log(rest_mass_density / reference_density_);
+  if (x <= 0.) {
+    return reference_pressure_ / reference_density_ /
+           (gamma_coefficients_[0] - 1.0) *
+           exp((gamma_coefficients_[0] - 1.0) * x);
+  }
+  const size_t n_points_epsilon = table_of_specific_energies_.size();
+  const double delta_x = x_max_ / (n_points_epsilon - 1.0);
+  double specific_energy = 0.0;
+  if (x >= x_max_) {
+    // Analytic integral at constant gamma_coefficient above x_max_
+    specific_energy = table_of_specific_energies_[n_points_epsilon - 1];
+    specific_energy += pressure_from_log_density(x_max_) / reference_density_ *
+                       exp(-x_max_) *
+                       (exp((gamma_of_x_max_ - 1.0) * (x - x_max_)) - 1.0) /
+                       (gamma_of_x_max_ - 1.0);
+    return specific_energy;
+  } else {
+    const auto table_index = static_cast<size_t>(floor(x / delta_x));
+    const double x0 = delta_x * table_index;
+    specific_energy = table_of_specific_energies_[table_index];
+    // Gaussian quadrature integral
+    for (size_t q = 0; q < number_of_quadrature_coefs_; q++) {
```

Comment:
In the future this could probably be factored out, but it's probably not worth it right now.